### ISSUE TITLE
new 16 bit api breaks old api

### ISSUE THREAD

**Issue by graphitemaster** [nothings/stb#398]
With the introduction of the new 16 bit API, changes were made which break the loading of existing PNGs.

The function
`stbi__load_and_postprocess_8bit`

Now assumes `ri.bits_per_channel` will be 8 or 16. However it's possible for the PNG loader to give this a different value than 8 or 16 in function `stbi__do_png` via `ri->bits_per_channel = p->depth;`

This results in this code being executed
```
   if (ri.bits_per_channel != 8) {
      STBI_ASSERT(ri.bits_per_channel == 16);
      result = stbi__convert_16_to_8((stbi__uint16 *) result, *x, *y, req_comp == 0 ? *comp : req_comp);
      ri.bits_per_channel = 8;
   }
```

Which then triggers the assertion and causes my application to crash. Removing the assertion also does not work because then it incorrectly converts as if it were 16 bits per channel.

I see two possible fixes here. Change `stbi__load_and_postprocess_8bit` to check for `> 8` instead of `!= 8` or change `stbi__do_png` to give a value of 8 for any bit depth < 16.

**Comment by nothings** [comment]
I imagine the correct fix is for stbi__do_png to return the *output* bits_per_channel, which is presumably to (as you say) give a value of 8 for any bit depth < 16, since <8-bit data is expanded to 8 bits-per-channel. It shouldn't be exposing the file-format bits-per-channel in the first place, that was a bug.

**Comment by graphitemaster** [comment]
Yeah that works and should be the fix, small patch provided below
```
static void *stbi__do_png(stbi__png *p, int *x, int *y, int *n, int req_comp, stbi__result_info *ri)
{
   void *result=NULL;
   if (req_comp < 0 || req_comp > 4) return stbi__errpuc("bad req_comp", "Internal error");
   if (stbi__parse_png_file(p, STBI__SCAN_load, req_comp)) {
-     ri->bits_per_channel = p->depth;
+     ri->bits_per_channel = p->depth < 16 ? 8 : 16;

**Comment by rygorous** [comment]
Fixed in commit 48710234f283ec106d429a28d53436133734d121.
